File: Question3.java
//Search in 2D matrix
//TC:- log(m)*log (n)
//SC: o(1)

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int low = 0;
        int high = m*n-1;

        while(low<=high) {
            int mid = low +(high - low)/2;
            int rowIdx = mid/n;
            int colIdx = mid%n;
            if(target == matrix[rowIdx][colIdx]) {
                return true;
            }
            else if(target < matrix[rowIdx][colIdx]) {
                high = mid-1;
            }else {
                low = mid+1;
            }
        }
        return false;

    }
}

File: Question_3.cpp
    
    int search(const ArrayReader& reader, int target) {
        int low = 0;
        while(reader.get(low)<=target){
            if(reader.get(low)==target) return low;
            low = 2*low+1;
        }
        int high = low;
        low = (low-1)/2;
        while(low<high){
            int mid = (low+high)/2;
            if(reader.get(mid)==target) return mid;
            else if(reader.get(mid)>target) high = mid-1;
            else low = mid+1;
        }
        if(reader.get(low)==target) return low;
        return -1;
    }